### DISCUSSION TITLE
How to generate lots of dynamic sitemaps when deploying to Vercel?

### DISCUSSION THREAD

**Discussion by techrome** [vercel/next.js#24339] [Help] [answered]
Hi guys, please help if anyone knows how to do this.
Suppose I have many blog pages (100 000 for example), then million of projects pages etc, so I have a total of 1 100 000 page URLs (for example). 
As we know, sitemap allows only a maximum of 50 000 URLs in a single file. So I will have to generate many sitemaps dynamically (and not on build time). 

So far I have made such a solution:
1. Get necessary urls for blog, then for projects, etc. 
2. Split them by 50000 chunks and then create files "1.xml", "2.xml" etc.
3. So now you have these sitemap files "/sitemaps/blog/1.xml", "/sitemaps/blog/2.xml", "/sitemaps/projects/1.xml" etc.
4. then create a single big sitemap file that combines all these sitemap file paths into a single `<sitemapindex>` tag and serve that through getServerSideProps with "application/xml". 

So far so good, but the problem arises when deploying to Vercel. Because I use "fs" module to write files to "/public" folder, Vercel doesn't allow to dynamically update/delete files inside public folder. 

So I have to find another method to upload files somewhere. I have tried using Dropbox, Google Cloud etc. but it seems so difficult there just to store files. I also tried using S3 but I think there's no support for removing folders. Maybe you guys have a better opinion how to do this or maybe I'm just dumb and going in a totally wrong direction? :D 
I would appreciate your help so much!

**Comment by gomflo** [comment]
You can try using `next-sitemap` exactly the dynamic/server-side feature, it can be helpful because your not doing it at build time.

https://github.com/iamvishnusankar/next-sitemap#generating-dynamicserver-side-sitemaps

**Reply by techrome** [reply]
Thanks for your reply! 
Yeah I know I can do that with `getServerSideProps`, but the problem is this page will be able to hold maximum of 50 000 entries for a sitemap, and if I make a dynamic routing like that for example `/sitemaps/blog/[number].js` and use `1.xml`, `2.xml` etc. then it will be very difficult to know what urls should start at `2.xml` for example and it just will be really messy. 
That's why I need to generate sitemap files so then I can access them whenever I want and they won't make a big load on a server.

Currently I'm looking at `neocities` free api for file storage I think it will work :D

**Comment by techrome** [comment, marked as answer]
**SOLUTION!** 
Surprisingly, the Neocities API did the trick and it was the easiest to integrate with 
(in contrast with Google Drive, Dropbox etc. 😄).

You can look at my (not fully ready yet, but working as expected) version of this on `sitemaps-test` branch: [https://github.com/techrome/my-portfolio/tree/sitemaps-test](https://github.com/techrome/my-portfolio/tree/sitemaps-test)

Basically, the whole process is as follows:
1. At build time, the `robots.txt` is generated by `next-sitemap`
2. There it also excludes my custom route `/dynamic-sitemap.xml` from static generation of sitemaps, and includes it as an additional sitemap.
3. So, whenever someone accesses it at `/dynamic-sitemap.xml`, getServerSideProps runs and returns `xml` with revalidate of some seconds (this works for Vercel so I can cache the result and not load the server unnecessary (like ISG works))
4. Server runs code that first removes (just to be safe) sitemaps folder from remote file storage (neocities.org in my example)
5. Then creates sitemap folder with the help of `memfs` (almost the same module as `fs` but creates files in memory, just to be able to send them via FormData)
6. Then creates each sitemap in memory and sends them to the file storage api.
7. Then removes in-memory folder of sitemaps (just for cleanup)
8. Then via some processing, it returns ready sitemap array which I can append to the xml file and that's it

I know for pages less than 10k-50k it seems like an overkill, but I don't like the idea of being limited so that's why I had to go to such lengths for making dynamic sitemaps :D

Maybe I forgot some details here, but when I remember I will post them here :)
Hope anyone finds this useful! Cheers
